driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -384,7 +309,32 @@ public void returnConnection(Connection connection) {
             if (connection.maxAvailableStreams() < minAllowedStreams) {
                 replaceConnection(connection);
             } else {
-                signalAvailableConnection();
+                dequeue(connection);
+            }
+        }
+    }
+
+    private void dequeue(Connection connection) {
+        while (!pendingBorrows.isEmpty()) {
+
+            // Try to acquire the right to use the connection
+            while (true) {
+                int inFlight = connection.inFlight.get();
+                if (inFlight >= Math.min(connection.maxAvailableStreams(), options().getMaxRequestsPerConnection(hostDistance))) {
+                    // Connection is full again, stop dequeuing
+                    return;
+                }
+                if (connection.inFlight.compareAndSet(inFlight, inFlight + 1))
```

Comment:
I tried to explain it in the comment above the `while` loop: a successful increment is the precondition that gives you the "right" to use the connection. I'll see how I can explain it better.